### ISSUE TITLE
bug: Tools aren't working in version 0.4.3

### ISSUE THREAD

**Issue by Simi5599** [open-webui/open-webui#7254]
# Bug Report

## Installation Method

Docker

## Environment

- **Open WebUI Version:** 0.4.3

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [x] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

When asking something to the model that requires a tool, the tool should activate

## Actual Behavior:

The tool never activates, even if enabled for the chat. I have downgraded to 0.4.2 and tools are working again.

## Reproduction Details

**Steps to Reproduce:**
0) Be sure that you have a tool that worked in version 0.4.2 (For example: "Run Code" by EtiennePerrot

1) Ask the lmm something

2) The LMM won't activate the tool

## Additional Information

Definitely can confirm that this is introduced in the 0.4.3 version.

**Comment by tjbck** [comment]
![image](https://github.com/user-attachments/assets/fbafb21d-719c-4a34-ae6b-ba8593a97b74)

Definitely works on my end, we might need more details here.

**Comment by Simi5599** [comment]
This is strange. Will update with a comment soon when I am in front of my laptop by attaching logs

**Comment by Simi5599** [comment]
So i have updates, the "Run code" tools works fine it was just a coincidence.
But the web search tool (not the one included by default in Open WebUI) does not work in any case. (in fact i did notice this issue by using the web search) 
By reading the Docker logs i see taht the error is related to the event_emitter.
Maybe this is a deprecation from Open WebUI?

**Browser Console Logs:**
Nothing relevant

**Docker Container Logs:**
I have got this error:

INFO  [open_webui.apps.openai.main] get_all_models()
/usr/local/lib/python3.11/site-packages/pydantic/main.py:1552: RuntimeWarning: fields may not start with an underscore, ignoring "__event_emitter__"
  warnings.warn(f'fields may not start with an underscore, ignoring "{f_name}"', RuntimeWarning)
ERROR [open_webui.main] Fields must not use names with leading underscores; e.g., use 'event_emitter__' instead of '__event_emitter__'.
Traceback (most recent call last):
  File "/app/backend/open_webui/main.py", line 679, in dispatch
    body, flags = await chat_completion_tools_handler(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/main.py", line 391, in chat_completion_tools_handler
    tools = get_tools(
            ^^^^^^^^^^
  File "/app/backend/open_webui/utils/tools.py", line 77, in get_tools
    "pydantic_model": function_to_pydantic_model(callable),
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/utils/tools.py", line 148, in function_to_pydantic_model
    return create_model(func.__name__, **field_defs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pydantic/main.py", line 1600, in create_model
    return meta(
           ^^^^^
  File "/usr/local/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 115, in __new__
    private_attributes = inspect_namespace(
                         ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 403, in inspect_namespace
    raise NameError(
NameError: Fields must not use names with leading underscores; e.g., use 'event_emitter__' instead of '__event_emitter__'.
generate_queries
gpt-4o-mini-2024-07-18
INFO:     151.57.200.225:0 - "POST /ws/socket.io/?EIO=4&transport=polling&t=PDLDqGJ&sid=7LZ6KPgpH6wotc46AAAC HTTP/1.1" 200 OK
INFO:     151.57.200.225:0 - "GET /ws/socket.io/?EIO=4&transport=polling&t=PDLDna3&sid=7LZ6KPgpH6wotc46AAAC HTTP/1.1" 200 OK

**Comment by tjbck** [comment]
Logs above aren't related. Perhaps @michaelpoluektov could chime in.

**Comment by Simi5599** [comment]
Could we do something like this?

    for name, param in parameters.items():
    # Rename parameters that starts with __
    if name.startswith('__'): 
        name = 'param_' + name[2:]  # Remove __ and add "param_"

**Comment by tjbck** [comment]
reserved params should NOT be added to the function spec, dropping reserved params is intended here.

**Comment by Simi5599** [comment]
Can confirm this works on my side and does not break tools.

![Screenshot 2024-11-22 205111](https://github.com/user-attachments/assets/8eb0252b-e968-4fd9-bb7a-0ffb72dbfd00)

If a simple PR is welcome i can do this without any issues :)

**Comment by michaelpoluektov** [comment]
I'm pretty sure everything works as intended here (apart from _maybe_ the fact that it's not failing fast enough, but that's a separate issue)

You should not be using reserved params (params starting with "__") in your tool definitions, unless they're part of OWUI (like `__event_emitter__`, `__event_call__` etc.)

If this isn't what's happening, feel free to attach your toolkit (or a dummy with the same type spec) and I'll look into it.

**Comment by Simi5599** [comment]
I am not, the only part of my code where i am using this kind of parameter is the following

``async def search_web(self, query: str, __event_emitter__: Callable[[dict], Any] = None)``

**Comment by michaelpoluektov** [comment]
It works for me, could you include your tool? Here's my type signature:

```python
    async def get_documents_by_id(
        self, ids: list[int], __event_emitter__: Callable[[dict], Awaitable] = None
    ) -> str:
```

**Comment by Simi5599** [comment]
The tool ha a lot of row, I can try to recreate an example when i come back to work.

**Comment by Simi5599** [comment]
Oh 😂 that's unexpected.

Couldn't we just filter out these kinds of parameters via the simple if?

**Comment by michaelpoluektov** [comment]
So simple fix for now: just remove:

```python
:param __event_emitter__:
```

from your docstring.

Otherwise I put up a PR: can you check that it works?

https://github.com/open-webui/open-webui/pull/7263

**Comment by Simi5599** [comment]
Can't check this at the moment, but I saw your PR, and it should work.

**Comment by tjbck** [comment]
0.4.4 will be released shortly!

**Comment by Simi5599** [comment]
Can confirm this was resolved with 0.4.4 !
